Crafting SecretTemplate how to guide. (#2249)
vmware-tanzu/tanzu-dev-portal: content/guides/tanzu-service-secret-sauce/index.md

There are a number of services available on Kubernetes, meaning in theory, developers have a plethora of options when it comes to what services they can deploy and consume.  In addition, Tanzu Application Platform (TAP) implements the [Kubernetes service binding specification](https://github.com/servicebinding/spec) allowing developers and DevOps personnel to exploit the power of service bindings and reduce the burden of configuration management in the context of consuming services.  The limitation of TAP service bindings is not grounded in the capabilities of the platform but instead by the small number of services that create configuration that is conformant with the Kubernetes service binding spec.  How can you transform a service’s configuration data into configuration that is compliant with the service binding specification?  A very powerful and clever option is through the use of [SecretTemplates](https://github.com/vmware-tanzu/carvel-secretgen-controller/blob/develop/docs/secret-template.md).

> In this guide, you will be introduced to the new _SecretTemplate_ custom resource, deploy an instance of a Redis service, aggregate the Redis configuration and secrets into a service binding compliant secret, and finally, build and deploy a trivial SpringBoot application that consumes the Redis service through the [TAP services toolkit](https://docs.vmware.com/en/Services-Toolkit-for-VMware-Tanzu-Application-Platform/0.6/svc-tlk/GUID-overview.html).  The examples in this guide are derived from a SecretTemplate [POC available on GitHub.](https://github.com/gm2552/redis-secret-template) 

Reviewer: ```suggestion
In this guide, you will be introduced to the new _SecretTemplate_ custom resource, deploy an instance of a Redis service, aggregate the Redis configuration and secrets into a service binding compliant secret, and finally, build and deploy a trivial Spring Boot application that consumes the Redis service through the [TAP services toolkit](https://docs.vmware.com/en/Services-Toolkit-for-VMware-Tanzu-Application-Platform/0.6/svc-tlk/GUID-overview.html).  The examples in this guide are derived from a SecretTemplate [POC available on GitHub.](https://github.com/gm2552/redis-secret-template) 
```

## Prerequisites

This guide assumes the following prerequisites are met:

* **Tanzu Cluster Essentials (TCE)** - This is a prerequisite of Tanzu Application Platform.  This guide uses resources that require TCE 1.2 or greater. 
* **Tanzu Application Platform (TAP)** – Installation of TAP 1.2.x or greater utilizing the “iterate” profile or other profile that has deployed out-of-the-box supply chains, out-of-the-box templates, services toolkit, and service bindings.
* **kubectl** – You will use this to manage Kubernetes resources. Installation instructions are[ here](https://kubernetes.io/docs/tasks/tools/).
* **Tanzu CLI** – You will use this to execute Tanzu specific operations. Install instructions are[ here](https://docs.vmware.com/en/VMware-Tanzu-Application-Platform/1.2/tap/GUID-install-tanzu-cli.html).
* **YTT** – You will use this to install a supply chain with user provided values. Install instructions are[ here](https://carvel.dev/ytt/docs/v0.42.0/install/).

> * You will need permissive access to the TAP Kubernetes cluster with kubectl and Tanzu cli tools

Reviewer: ```suggestion
* You will need permissive access to the TAP Kubernetes cluster with kubectl and Tanzu CLI tools
```

```sh
kubectl create ns service-instances

kubectl apply -f https://raw.githubusercontent.com/RedisLabs/redis-enterprise-k8s-docs/v6.2.10-45/bundle.yaml -n service-instances
```

Validate the operator is installed with the following command:
```sh
kubectl get pods -l name=redis-enterprise-operator -n service-instances
```

You should see an output similar to the following:
```text
NAME                                         READY   STATUS    RESTARTS   AGE
redis-enterprise-operator-74849d8c69-pnqvq   2/2     Running   0          39s
```

Now that the operator is deployed, you can create a Redis cluster and database.  To deploy the cluster and database, run the command below.  The configuration referenced in the following command also deploys a SecretTemplate resource to convert the subsequent Redis configuration into a service binding compliant secret:
```sh
ytt -f https://raw.githubusercontent.com/gm2552/redis-secret-template/main/templates/redisEnterpriseClusterOperator/redisEnterpriseClusterTemplate.yaml -v service_namespace=service-instances -v instance_name=redis-test | kubectl apply -f-
```

> This will create a cluster and database named _redis-test_.  Once the cluster is ready, its configuration is written to a secret and service resource.

Reviewer: ```suggestion
This will create a cluster and database named `redis-test`.  Once the cluster is ready, its configuration is written to a secret and service resource.
```
